core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -370,27 +368,54 @@ public ShardStartedClusterStateTaskExecutor(AllocationService allocationService,
         }
 
         @Override
-        public String describeTasks(List<ShardRoutingEntry> tasks) {
-            return tasks.stream().map(entry -> entry.getShardRouting().toString()).reduce((s1, s2) -> s1 + ", " + s2).orElse("");
-        }
-
-        @Override
-        public BatchResult<ShardRoutingEntry> execute(ClusterState currentState, List<ShardRoutingEntry> tasks) throws Exception {
-            BatchResult.Builder<ShardRoutingEntry> builder = BatchResult.builder();
-            List<ShardRouting> shardRoutingsToBeApplied = new ArrayList<>(tasks.size());
-            for (ShardRoutingEntry task : tasks) {
-                shardRoutingsToBeApplied.add(task.shardRouting);
+        public BatchResult<ShardEntry> execute(ClusterState currentState, List<ShardEntry> tasks) throws Exception {
+            BatchResult.Builder<ShardEntry> builder = BatchResult.builder();
+            List<ShardEntry> validTasks = new ArrayList<>();
+            List<ShardRouting> validShardRoutings = new ArrayList<>(tasks.size());
+            Set<ShardRouting> seenValidShardRoutings = new HashSet<>();
+            for (ShardEntry task : tasks) {
+                assert task.primaryTerm == 0L : "shard is only started by itself";
+
+                ShardRouting matched = currentState.getRoutingTable().getByAllocationId(task.shardId, task.allocationId);
+                if (matched == null) {
+                    // tasks that correspond to non-existent shards are marked as successful
```

Comment:
this is a bit funky - can you add to the comment way it's ok (old stale requests that might be still in flight). My guess is thai is done partially because we resend shard started on every cluster state publishing that doesn't contain the shard as started so we have many of these in flight.   I think we will revisit this at some point - we shoudl have all the "channel failures" like master loss covered already.